[WebGL] Avoid using glReadPixels to paint WebGL canvas into 2D canvas

djg opened the pull request:
#### 50ae4e81ccce7b0c1ebd530b24a88ed25ba82caf
<pre>
[WebGL] Avoid using glReadPixels to paint WebGL canvas into 2D canvas
<a href="https://bugs.webkit.org/show_bug.cgi?id=252795">https://bugs.webkit.org/show_bug.cgi?id=252795</a>
rdar://87173127

Reviewed by Kimmo Kinnunen.

Instead of reading pixels from the current drawing buffer using glReadPixels,
the IOSurface backing the drawing buffer is wrapped in a CGImage for the
duration of the synchronous call to paintRenderingResultsToCanvas.

To ensure that the WebGL rendering is present in the drawing buffer, any pending
GL operations are executed to update rendering before creating a CGImageRef
wrapper around WebGL&apos;s IOSurface.

* Source/WebCore/platform/graphics/GraphicsContextGL.cpp:
(WebCore::GraphicsContextGL::paintToCanvas):

Platform dependent code has been extracted into
createNativeImageFromPixelBuffer, allowing paintToCanvas implementations to be
unified.

* Source/WebCore/platform/graphics/GraphicsContextGL.h:

Split paintToCanvas(..., Ref&lt;PixelBuffer&gt;&amp;&amp;, ...) into
createNativeImageFromPixelBuffer(Ref&lt;PixelBufer&gt;&amp;&amp;, ...) and
paintToCanvas(NativeImage&amp;, ...). This allows the painting of any NativeImage to
canvas, not just PixelBuffer sources and provides a convenience function to
obtain a NativeImage from a PixelBuffer.

* Source/WebCore/platform/graphics/angle/GraphicsContextGLANGLE.cpp:
(WebCore::GraphicsContextGLANGLE::prepareTexture):

Since we&apos;re now calling prepareTexture() in a difference context remove ASSERT
that layer isn&apos;t composited.

* Source/WebCore/platform/graphics/angle/GraphicsContextGLANGLE.h:
* Source/WebCore/platform/graphics/cairo/GraphicsContextGLCairo.cpp:
(WebCore::GraphicsContextGL::createNativeImageFromPixelBuffer):

Helper function to create NativeImage from PixelBuffer.

(WebCore::GraphicsContextGL::paintToCanvas): Deleted.
* Source/WebCore/platform/graphics/cg/GraphicsContextGLCG.cpp:
(WebCore::GraphicsContextGL::createNativeImageFromPixelBuffer):

Helper function to create NativeImage from PixelBuffer.

(WebCore::GraphicsContextGL::paintToCanvas): Deleted.
 Source/WebCore/platform/graphics/cocoa/GraphicsContextGLCocoa.h:
* Source/WebCore/platform/graphics/cocoa/GraphicsContextGLCocoa.mm:
(WebCore::GraphicsContextGLCocoa::paintRenderingResultsToCanvas):
(WebCore::GraphicsContextGLCocoa::paintCompositedResultsToCanvas):

Defer implementation using helper functions and paintToCanvas(NativeImage&amp;, ...)

(WebCore::GraphicsContextGLCocoa::withDrawingBufferAsNativeImage):
(WebCore::GraphicsContextGLCocoa::withDisplayBufferAsNativeImage):

Helper methods to aid borrowing WebGL&apos;s IOSurface-backed drawing and display
buffers as NativeImage, taking care of conversion, context creation, flushing,
etc. WebGL contexts not using premultiplied alpha still use the glReadPixels
path because IOSurfaces without premultipled alpha are unsupported by
CoreGraphics. The image is passed by reference to stop the function from hanging
onto the image.

* Source/WebCore/platform/graphics/cocoa/GraphicsContextGLIOSurfaceSwapChain.h:
* Source/WebKit/GPUProcess/graphics/RemoteGraphicsContextGL.cpp:
(WebKit::RemoteGraphicsContextGL::paintRenderingResultsToCanvasWithQualifiedIdentifier):
(WebKit::RemoteGraphicsContextGL::paintCompositedResultsToCanvasWithQualifiedIdentifier):

Defer implementation using helper functions and paintNativeImageToImageBuffer()

(WebKit::RemoteGraphicsContextGL::paintNativeImageToImageBuffer):
(WebKit::RemoteGraphicsContextGL::paintPixelBufferToImageBuffer): Deleted.

Painting to ImageBuffer is only supported via NativeImage. To paint a pixel
buffer, first convert the PixelBuffer to NativeImage using
createNativeImageFromPixelBuffer().

* Source/WebKit/GPUProcess/graphics/RemoteGraphicsContextGL.h:

Canonical link: <a href="https://commits.webkit.org/262518@main">https://commits.webkit.org/262518@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/401516891037ccd92b1a64ed55f8af876be8632d

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/1654 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/1685 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/1743 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/2577 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/1777 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/1635 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/1752 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/1749 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/1553 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/1669 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/1496 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/1490 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/2413 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/1493 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/1476 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/1394 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/1514 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/1518 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/2579 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/1547 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/1378 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/1465 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/1483 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/432 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/1611 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

kkinnunen-apple commented on Source/WebCore/platform/graphics/cg/GraphicsContextGLCG.cpp:
```diff
@@ -531,7 +535,16 @@ void GraphicsContextGL::paintToCanvas(const GraphicsContextGLAttributes& sourceC
+    if (canvasSize.isEmpty())
+        return;
+
+    auto image = NativeImage::create(WTFMove(platformImage));
```
I think you could revert the NativeImage -> PlatformImagePtr changes. E.g. if only possible, use NativeImage, a real abstraction, as opposed to typedef hack.

djg replied:
Nice. This cleans up the duplication of paintToCanvas between Cairo and CG.

kkinnunen-apple commented on Source/WebCore/platform/graphics/cocoa/GraphicsContextGLCocoa.h:
```diff
@@ -103,6 +109,7 @@ class WEBCORE_EXPORT GraphicsContextGLCocoa : public GraphicsContextGLANGLE {
     bool reshapeDisplayBufferBacking() final;
     bool allocateAndBindDisplayBufferBacking();
     bool bindDisplayBufferBacking(std::unique_ptr<IOSurface> backing, void* pbuffer);
+    std::pair<std::unique_ptr<IOSurface>, void*> unbindDisplayBufferBacking();
```
leftover?

kkinnunen-apple commented on Source/WebCore/platform/graphics/AlphaMode.h:
```diff
@@ -0,0 +1,36 @@
+
+namespace WebCore {
+
+enum class AlphaMode : uint8_t {
```
This could be synched up with WebCore::AlphaPremultiplication. At least:
- have same nomenclature (Unmultiplied vs Unpremultiplied)

Alternatively, perhaps preferred, would be to investigate fixing `GraphicsContextGLANGLE::readPixelsForPaintResults()`.
So there we would indicate AlphaMode::None by setting PixelFormat::RGBX8.
And premul/unpremul with AlphaPremultiplication.

kkinnunen-apple commented on Source/WebCore/platform/graphics/cocoa/GraphicsContextGLCocoa.mm:
```diff
@@ -906,6 +907,90 @@ static bool needsEAGLOnMac()
+    // ImageBufferIOSurfaceBackend::prepareForExternalRead(), but that's
+    // private.
+    CGContextFlush(cgContext.get());
+    WTFLogAlways("GraphicsContextGLCocoa(%p)::withDisplayBufferAsPlatformImage(...) <<<", this);
```
remove debug print

kkinnunen-apple commented on Source/WebCore/platform/graphics/cocoa/GraphicsContextGLCocoa.mm:
```diff
@@ -906,6 +907,90 @@ static bool needsEAGLOnMac()
+    // creating a copy?
+    func(displayBuffer.surface->createImage(cgContext.get()));
+
+    // FIXME: Should we be interacting with ImageBufferIOSurfaceBackend
```
so here you don't need to flush the context used for the **source image**, as  that does not have any drawing.
Instead, you could do `CGContextFillRect(cgContext.get(), CGRect { });`.
This would ensure that if the func() hangs on to the image, we would not continue updating the image behind the func()s back.

kkinnunen-apple commented on Source/WebCore/platform/graphics/cocoa/GraphicsContextGLCocoa.mm:
```diff
@@ -906,6 +907,90 @@ static bool needsEAGLOnMac()
+    if (displayBuffer.surface->size() != getInternalFramebufferSize())
+        return;
+
+    // FIXME: Should a displayID be passed here?
```
No need, you can remove this.

kkinnunen-apple commented on Source/WebCore/platform/graphics/cocoa/GraphicsContextGLCocoa.mm:
```diff
@@ -906,6 +907,90 @@ static bool needsEAGLOnMac()
+    if (!cgContext)
+        return;
+
+    // FIXME: Should this be CGIOSurfaceContextCreateImageReference to avoid
```
No need at the moment, you can remove this. There's no difference in this invocation sequence.
The create image is better at the moment, as we "don't know" if the func() will hang on to the image or not.

kkinnunen-apple reviewed (commented):
Very cool stuff!
It would be very helpful if you could submit the ANGLE parts directly to ANGLE, we should be able to get it reviewed there and merged back with quite minimal overhead.
Feel free to keep the hunks in this patch while you work on the WebKit side, but if you can start the upstreaming in parallel, that would help a lot!

kkinnunen-apple commented on Source/WebCore/platform/graphics/cocoa/GraphicsContextGLCocoa.mm:
```diff
@@ -906,6 +907,90 @@ static bool needsEAGLOnMac()
+            // FIXME: Should this be CGIOSurfaceContextCreateImageReference to avoid
+            // creating a copy?
+
+            auto drawingImage = m_displayBufferBacking->createImage(cgContext.get());
```
add this: `CGImageSetCachingFlags(image.get(), kCGImageCachingTransient);`

kkinnunen-apple commented on Source/WebCore/platform/graphics/cocoa/GraphicsContextGLCocoa.mm:
```diff
@@ -906,6 +907,90 @@ static bool needsEAGLOnMac()
+
+    // FIXME: Should this be CGIOSurfaceContextCreateImageReference to avoid
+    // creating a copy?
+    func(displayBuffer.surface->createImage(cgContext.get()));
```
add this: `CGImageSetCachingFlags(image.get(), kCGImageCachingTransient);`

kkinnunen-apple commented on Source/WebKit/GPUProcess/graphics/RemoteGraphicsContextGL.cpp:
```diff
@@ -219,7 +219,9 @@ void RemoteGraphicsContextGL::paintRenderingResultsToCanvas(WebCore::RenderingRe
     assertIsCurrent(workQueue());
-    paintPixelBufferToImageBuffer(m_context->readRenderingResultsForPainting(), imageBuffer, WTFMove(completionHandler));
+    m_context->withDrawingBufferAsPlatformImage([&](PlatformImagePtr&& image) {
+        paintPlatformImageToImageBuffer(WTFMove(image), imageBuffer, WTFMove(completionHandler));
```
So here you could add:

```
// We know that the image might be updated afterwards, so flush the drawing so that read back does not
occur.
imageBuffer->flushDrawingContext();
```

djg replied:
This flushDrawingContext is happening inside paintPlatformImageToImageBuffer. Added the comment to that location.

kkinnunen-apple commented on Source/WebKit/GPUProcess/graphics/RemoteGraphicsContextGL.cpp:
```diff
@@ -232,7 +234,9 @@ void RemoteGraphicsContextGL::paintCompositedResultsToCanvas(WebCore::RenderingR
 {
     assertIsCurrent(workQueue());
-    paintPixelBufferToImageBuffer(m_context->readCompositedResultsForPainting(), imageBuffer, WTFMove(completionHandler));
+    m_context->withDisplayBufferAsPlatformImage([&](PlatformImagePtr&& image) {
```
```
// We know that the image might be updated afterwards, so flush the drawing so that read back does not
occur.
imageBuffer->flushDrawingContext();
```
(this is not so self-evident at the moment for the composited image, but it's a good starting point anyway.

kkinnunen-apple commented on Source/WebCore/platform/graphics/cocoa/GraphicsContextGLCocoa.mm:
```diff
@@ -906,6 +907,90 @@ static bool needsEAGLOnMac()
+        // read the image into a PixelBuffer which can be we used to create a
+        // CGImage.
+        //
+        // FIXME: Can the IOSurface be read into a buffer to avoid the read back via GL?
```
So the IOSurface can be accessed via 
```
IOSurface::Locker locker { m_displayBufferBacking->surface() };
```
However:
-  the construction of the pixel buffer and data acquisition is bundled in paintRenderingResultsToPixelBuffer
- IOSurface is in BGRA where as the pixel buffer might need to be in RGBA
- IOSurface has row length != 4*width where as pixel buffer might need to be tightly packed
- Some other aspects

So it might be simpler to not do this in this patch.

kkinnunen-apple reviewed (approved):
Nice!

kkinnunen-apple commented on Source/WebCore/platform/graphics/cocoa/GraphicsContextGLCocoa.mm:
```diff
@@ -906,6 +907,78 @@ static bool needsEAGLOnMac()
+
+    if (contextAttributes().premultipliedAlpha) {
+        // Use the IOSurface backed image directly
+        EGL_WaitUntilWorkScheduledANGLE(platformDisplay());
```
In fact, this should be call to GraphicsContextGLCocoa::waitUntilWorkScheduled()

```
void GraphicsContextGLCocoa::waitUntilWorkScheduled() {
   if (is metal) {
           EGL_WaitUntilWorkScheduledANGLE(platformDisplay());
  } else {
         GL_Flush();
  }
}

```

djg committed:
[WebGL] Avoid using glReadPixels to paint WebGL canvas into 2D canvas
https://bugs.webkit.org/show_bug.cgi?id=252795
rdar://87173127

Reviewed by Kimmo Kinnunen.

Instead of reading pixels from the current drawing buffer using glReadPixels,
the IOSurface backing the drawing buffer is wrapped in a CGImage for the
duration of the synchronous call to paintRenderingResultsToCanvas.

To ensure that the WebGL rendering is present in the drawing buffer, any pending
GL operations are executed to update rendering before creating a CGImageRef
wrapper around WebGL's IOSurface.

* Source/WebCore/platform/graphics/GraphicsContextGL.cpp:
(WebCore::GraphicsContextGL::paintToCanvas):

Platform dependent code has been extracted into
createNativeImageFromPixelBuffer, allowing paintToCanvas implementations to be
unified.

* Source/WebCore/platform/graphics/GraphicsContextGL.h:

Split paintToCanvas(..., Ref<PixelBuffer>&&, ...) into
createNativeImageFromPixelBuffer(Ref<PixelBufer>&&, ...) and
paintToCanvas(NativeImage&, ...). This allows the painting of any NativeImage to
canvas, not just PixelBuffer sources and provides a convenience function to
obtain a NativeImage from a PixelBuffer.

* Source/WebCore/platform/graphics/angle/GraphicsContextGLANGLE.cpp:
(WebCore::GraphicsContextGLANGLE::prepareTexture):

Since we're now calling prepareTexture() in a difference context remove ASSERT
that layer isn't composited.

* Source/WebCore/platform/graphics/angle/GraphicsContextGLANGLE.h:
* Source/WebCore/platform/graphics/cairo/GraphicsContextGLCairo.cpp:
(WebCore::GraphicsContextGL::createNativeImageFromPixelBuffer):

Helper function to create NativeImage from PixelBuffer.

(WebCore::GraphicsContextGL::paintToCanvas): Deleted.
* Source/WebCore/platform/graphics/cg/GraphicsContextGLCG.cpp:
(WebCore::GraphicsContextGL::createNativeImageFromPixelBuffer):

Helper function to create NativeImage from PixelBuffer.

(WebCore::GraphicsContextGL::paintToCanvas): Deleted.
 Source/WebCore/platform/graphics/cocoa/GraphicsContextGLCocoa.h:
* Source/WebCore/platform/graphics/cocoa/GraphicsContextGLCocoa.mm:
(WebCore::GraphicsContextGLCocoa::paintRenderingResultsToCanvas):
(WebCore::GraphicsContextGLCocoa::paintCompositedResultsToCanvas):

Defer implementation using helper functions and paintToCanvas(NativeImage&, ...)

(WebCore::GraphicsContextGLCocoa::withDrawingBufferAsNativeImage):
(WebCore::GraphicsContextGLCocoa::withDisplayBufferAsNativeImage):

Helper methods to aid borrowing WebGL's IOSurface-backed drawing and display
buffers as NativeImage, taking care of conversion, context creation, flushing,
etc. WebGL contexts not using premultiplied alpha still use the glReadPixels
path because IOSurfaces without premultipled alpha are unsupported by
CoreGraphics. The image is passed by reference to stop the function from hanging
onto the image.

* Source/WebCore/platform/graphics/cocoa/GraphicsContextGLIOSurfaceSwapChain.h:
* Source/WebKit/GPUProcess/graphics/RemoteGraphicsContextGL.cpp:
(WebKit::RemoteGraphicsContextGL::paintRenderingResultsToCanvasWithQualifiedIdentifier):
(WebKit::RemoteGraphicsContextGL::paintCompositedResultsToCanvasWithQualifiedIdentifier):

Defer implementation using helper functions and paintNativeImageToImageBuffer()

(WebKit::RemoteGraphicsContextGL::paintNativeImageToImageBuffer):
(WebKit::RemoteGraphicsContextGL::paintPixelBufferToImageBuffer): Deleted.

Painting to ImageBuffer is only supported via NativeImage. To paint a pixel
buffer, first convert the PixelBuffer to NativeImage using
createNativeImageFromPixelBuffer().

* Source/WebKit/GPUProcess/graphics/RemoteGraphicsContextGL.h:

Canonical link: https://commits.webkit.org/262518@main
